File: gcj2014R1AA.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

static auto solve = []()
{
    int N,L; std::cin >> N >> L;
    std::vector<unsigned long long> o(N,0ULL);
    std::vector<unsigned long long> d(N,0ULL);
    for(auto& v: o)
    {
        std::string s; std::cin >> s;
        for(auto c: s) v<<=1, v|=c!='0';
    }
    for(auto& v: d)
    {
        std::string s; std::cin >> s;
        for(auto c: s) v<<=1, v|=c!='0';
    }
    std::sort(d.begin(),d.end());
    const int INF = L+1;
    int n=INF;
    for(auto v: o)
    {
        auto t = o;
        const auto b = d[0]^v;
        for(auto& v: t) v^=b;
        std::sort(t.begin(),t.end());
        if(t==d) n=std::min(n,std::__pop_count(b));
    }
    if(n==INF) return std::string("NOT POSSIBLE");
    return std::to_string(n);
};

int main()
{
    std::cin.tie(0);
    std::ios_base::sync_with_stdio(false);
    int T; std::cin >> T;
    for (int i=1; i<=T; ++i)
        std::cout << "Case #" << i << ": " << solve() << std::endl;
}

File: gcj2014R1AA.hs
#!/usr/bin/env runghc
import Data.List
import Data.List.Split
import Data.Bits
main = getContents >>= mapM_ putStrLn . zipWith caseno [1..] . solve . lines
caseno i r = "Case #"++show i++": "++r
solve (c:cs) = map (toString . answer) . take (read c) . analyze $ cs ::[String]
analyze = map (map (map s2b . words) . tail) . chunksOf 3 :: [String]->[[[Integer]]]
  where s2b = foldl' (\b c-> shiftL b 1 + if c=='0' then 0 else 1) 0
toString a = if a/=inf then show a else "NOT POSSIBLE"
answer [os,ds] = minimum . map f $ os where
  ds' = sort ds
  f o = if ds' == (sort . map (xor b) $ os) then count b else inf
    where b = o `xor` head ds'
count 0 = 0
count b = 1 + count (b .&. (b-1))
inf = maxBound::Int
